src/hevm/src/EVM/Expr.hs

```diff
@@ -392,8 +392,18 @@ instance Monoid (Expr Buf) where
 -- | Removes any irrelevant writes when reading from a buffer
 simplifyReads :: Expr a -> Expr a
 simplifyReads = \case
-  ReadWord (Lit idx) b -> ReadWord (Lit idx) (stripWrites idx (idx + 31) b)
-  ReadByte (Lit idx) b -> ReadByte (Lit idx) (stripWrites idx idx b)
+  ReadWord (Lit idx) b ->
+    let a = ReadWord (Lit idx) (stripWrites idx (idx + 31) b)
+    in case a of
+      (ReadWord (Lit idx2) (ConcreteBuf buf)) ->
+        let unpacked = Data.List.take 32 $ Data.List.drop (fromIntegral $ toInteger idx2) $ BS.unpack buf
+        in Lit (W256 $ word256 $ BS.pack unpacked)
+      x -> x
+  ReadByte (Lit idx) b ->
+    let a = ReadByte (Lit idx) (stripWrites idx idx b)
+    in case a of
+      (ReadByte (Lit idx2) (ConcreteBuf buf)) -> LitByte (BS.unpack buf !! (fromIntegral $ toInteger idx2))
+      x -> x
```

Comment:
This simplification of reading bytes and words from concrete buffs is also achieved by `readBytes` and `readWords`. Maybe we could apply these instead?

Reply:
This has been moved to the things you recommended, thanks!